jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/ConfigurableRateLimiterKt.kt

```diff
@@ -0,0 +1,82 @@
+package ru.fix.stdlib.ratelimiter
+
+import io.github.resilience4j.ratelimiter.RateLimiterConfig
+import io.github.resilience4j.ratelimiter.internal.AtomicRateLimiter
+import kotlinx.coroutines.delay
+import java.time.Duration
+import java.time.temporal.TemporalUnit
+import java.util.concurrent.RejectedExecutionException
+import java.util.concurrent.TimeUnit
+
+class ConfigurableRateLimiterKt(name: String, permitsPerSecond: Int) : RateLimiterKt {
+
+    @Volatile
+    private var rateLimiter: io.github.resilience4j.ratelimiter.RateLimiter
+
+    init {
+        // AtomicRateLimiter does not distribute events inside refresh period.
+        // It releases all available permits immediately on interval start if there is a demand for them.
+        // To distribute event in our interval of 1 second we divide it
+        // into chunks so that 1 chunk of time is limited to 1 permit.
+
+        // AtomicRateLimiter does not distribute events inside refresh period.
+        // It releases all available permits immediately on interval start if there is a demand for them.
+        // To distribute event in our interval of 1 second we divide it
+        // into chunks so that 1 chunk of time is limited to 1 permit.
+        rateLimiter = AtomicRateLimiter(
+                name,
+                RateLimiterConfig.custom()
+                        .limitForPeriod(1)
+                        .limitRefreshPeriod(Duration.ofNanos((1000000000 / permitsPerSecond).toLong()))
+                        .build()
+        )
+    }
+
+    override fun close() {
+        // Nothing to do
+    }
+
+    override suspend fun getRate(): Double {
+        val refreshPeriodInNanos = rateLimiter.rateLimiterConfig.limitRefreshPeriodInNanos
+
+        return (1000000000 / refreshPeriodInNanos).toDouble()
+    }
+
+    override suspend fun tryAcquire(): Boolean {
+        return rateLimiter.acquirePermission(Duration.ZERO)
+    }
+
+    override suspend fun tryAcquire(timeout: Long, unit: TemporalUnit?): Boolean {
+        return rateLimiter.acquirePermission(Duration.of(timeout, unit))
+    }
+
+    override suspend fun updateRate(permitsPerSecond: Int) {
+        val name = rateLimiter.name
+
+        // We can update rate limiter at any time because limits for period is always set to 1.
+        // So at most we exceed limit by 1
+
+        // We can update rate limiter at any time because limits for period is always set to 1.
+        // So at most we exceed limit by 1
+        rateLimiter = AtomicRateLimiter(
+                name,
+                RateLimiterConfig.custom()
+                        .limitForPeriod(1)
+                        .limitRefreshPeriod(Duration.ofNanos((1000000000 / permitsPerSecond).toLong()))
+                        .build()
+        )
+    }
+
+    private suspend fun io.github.resilience4j.ratelimiter.RateLimiter.acquirePermission(timeout: Duration): Boolean {
+        val waitTimeNs = this.reservePermission(timeout)
```

Comment:
1. Am I correct, that you have implemented suspendable rate limiter permission logic await by yourself to do not update resilience4j version in jfix-stdlib? I am ok with it for now - just to be sure
2. When you specify timeout greater than zero, you make `reservePermission` method to block.

Reply:
We got resilience4j-ratelimiter of version 0.13.0. For resilience4j-kotlin we need a higher version, 1.7.0 for example. But if we update resilience4j-ratelimiter - it will break jave implementation of rate limiter.
Block of code taken from resilience4j-kotlin:
https://github.com/resilience4j/resilience4j/blob/ac71bf8815439f7b4b1d32df7611733f30bec6d9/resilience4j-kotlin/src/main/kotlin/io/github/resilience4j/kotlin/ratelimiter/RateLimiter.kt#L64